### DISCUSSION TITLE
Should Attempt even have a type parameter?

### DISCUSSION THREAD

**Discussion by sleberknight** [kiwiproject/retrying-again#28] [General]
We changed the raw usages of `Attempt` to use a type parameter in #23 , though they all ended up being `Attempt<?>` which is non-ideal at best.

Can we do better than this? 

Should `Attempt` even have a type parameter at the class level?

_Client code_ doesn't directly create `Attempt` instances and only gets them from `RetryException#getLastFailedAttempt`, which itself only occurs when all attempts of a `Retryer` fail. (Obviously inside the `Retryer` where the `Attempt` objects are instantiated, the full type information is available, but once it leaves via a `RetryerException` that information is currently lost.)

The _client code_ calling `RetryException#getLastFailedAttempt` cannot infer the generic type of the returned `Attempt` and so must use `Attempt<?>` (or use `var` but that's effectively the same thing since the explicit type is still `Attempt<?>`).

With explicit types shown for clarity you end up with code like this:

```java
Attempt<?> lastFailedAttempt = retryException.getLastFailedAttempt();
Object result = lastFailedAttempt.getResult();
```

And this forces you to cast the result to a type that _you_ (probably) know but which the compiler is unable to infer:

```java
var lastFailedAttempt = retryException.getLastFailedAttempt();
var result = (Long) lastFailedAttempt.getResult();
```

This could obviously result in other problems such as `ClassCastException`.

Some options would be to:

1. Do nothing
2. Remove the type parameter from `Attempt` so that we don't need to carry the `<?>` all over the place (which means the result is just `Object`)

The ideal option would be for the Attempt to carry the expected result type along with it such that if code only has a `RetryException`, it will be able to infer the type. I am not sure whether this is easily doable with the current design.

Any thoughts?

**Comment by chrisrohr** [comment]
Does adding the type to the class provide with your Option 3 where you can infer the type on the result?

**Reply by sleberknight** [reply]
`Attempt` is defined thusly: 😂

```java
public class Attempt<T> {

    private final T result;

   // ...
}
```

Which works just fine if you are the one who creates the `Attempt`, as is the case within `Retryer#call`.

But you cannot parameterize classes that extend `Throwable` (_Generic class may not extend 'java.lang.Throwable')_ and while you can do this in the constructor:

```java
<T> RetryException(@Nonnull Attempt<T> attempt) { ... }
```

you cannot then declare the internal field as `Attempt<T>` - it won't compile since you are not able to put a type parameter on the type itself. So that leaves you with `Attempt<?>`.

Now, maybe there is just something I am missing or simply don't know how to make this work in this situation (i.e. an Exception class)...

**Reply by chrisrohr** [reply]
Then I guess we will just have to fall back to Object in this case and people will have to cast

**Comment by sleberknight** [comment]
There's not much (good) we can do to "fix" this, so I have documented the restriction in #222 and am considering this as "closed".

**Reply by sleberknight** [reply]
Another "fix" is to add a method to `RetryException` to get the last failed `Attempt` but provide the expected type as an argument. See #224
